source/extensions/filters/http/ext_authz/ext_authz.cc

```diff
@@ -102,20 +94,53 @@ void Filter::onComplete(Filters::Common::ExtAuthz::CheckStatus status) {
     break;
   }
 
+  ENVOY_STREAM_LOG(trace, "Ext_authz received status code {}", *callbacks_,
+                   enumToInt(response->status_code));
+
   // We fail open/fail close based of filter config
   // if there is an error contacting the service.
-  if (status == CheckStatus::Denied ||
-      (status == CheckStatus::Error && !config_->failureModeAllow())) {
-    callbacks_->sendLocalReply(Http::Code::Forbidden, "", nullptr);
+  if (response->status == CheckStatus::Denied ||
+      (response->status == CheckStatus::Error && !config_->failureModeAllow())) {
+    ENVOY_STREAM_LOG(debug, "Ext_authz rejected the request", *callbacks_);
+    callbacks_->sendLocalReply(
+        response->status_code, response->body,
+        [& authz_headers = response->headers_to_add,
+         &callbacks = *callbacks_ ](Http::HeaderMap & response_headers)
+            ->void {
+              for (const auto& header : authz_headers) {
+                ENVOY_STREAM_LOG(trace, "Ext_authz rejected response header '{}':'{}'", callbacks,
+                                 header.first.get(), header.second);
+                response_headers.addReferenceKey(header.first, header.second);
+              }
+            });
     callbacks_->requestInfo().setResponseFlag(
         RequestInfo::ResponseFlag::UnauthorizedExternalService);
   } else {
+    ENVOY_STREAM_LOG(debug, "Ext_authz accepted the request", *callbacks_);
     // Let the filter chain continue.
     filter_return_ = FilterReturn::ContinueDecoding;
-    if (config_->failureModeAllow() && status == CheckStatus::Error) {
+    if (config_->failureModeAllow() && response->status == CheckStatus::Error) {
       // Status is Error and yet we are allowing the request. Click a counter.
       cluster_->statsScope().counter("ext_authz.failure_mode_allowed").inc();
     }
+    // Only send headers if the response is ok.
+    if (response->status == CheckStatus::OK) {
+      for (const auto& header : response->headers_to_add) {
+        request_headers_->setReferenceKey(header.first, header.second);
+        ENVOY_STREAM_LOG(trace, "Ext_authz ok response added header '{}':'{}'", *callbacks_,
+                         header.first.get(), header.second);
+      }
+      for (const auto& header : response->headers_to_append) {
+        Http::HeaderEntry* header_to_modify = request_headers_->get(header.first);
+        if (header_to_modify) {
+          header_to_modify->value().append(", ", 2);
+          header_to_modify->value().append(header.second.c_str(), header.second.size());
```

Comment:
Should we just have an `append` method on `HeaderEntry`? You could also use the `HeaderMap::add*` methods, which add an additional header, which I think (?) is equivalent in the general case.